File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    static class Location {
        int x;
        int y;
        int cost;
        
        public Location(int x, int y, int cost) {
            this.x = x;
            this.y = y;
            this.cost = cost;
        }
    }
    
    static int n, m;
    static boolean[][] visited;
    static final int[] dx = { 0, 0, -1, 1 };
    static final int[] dy = { 1, -1, 0, 0 };
    
    public int solution(int[][] maps) {
        
        n = maps.length;
        m = maps[0].length;
        
        visited = new boolean[n][m];
        
        int answer = 0;
        answer = bfs(maps);
        return answer;
    }
    
    static int bfs(int[][] graph) {
        Deque<Location> queue = new ArrayDeque<>();
        queue.add(new Location(0,0,1));
        visited[0][0] = true;
        
        while (!queue.isEmpty()) {
            
            Location cur = queue.poll();
            int x = cur.x;
            int y = cur.y;
            int cost = cur.cost;
            
            if (x == n-1 && y == m-1) {
                return cost;
            }
            
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
                    continue;
                }
                
                if (graph[nx][ny] == 1 && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    queue.add(new Location(nx, ny, cost+1));
                }
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
using namespace std;

int dx[] = {1,-1,0,0};
int dy[] = {0,0,1,-1};

int solution(vector<vector<int>> maps)
{
    int answer = 0;
    int n = maps.size();
    int m = maps[0].size();
    
    deque<vector<int>> queue;
    queue.push_back({0,0,1});
    
    while (!queue.empty()) {
        auto cur = queue.front();
        queue.pop_front();
        int x = cur[0];
        int y = cur[1];
        int cost = cur[2];
        
        if (x == n-1 && y == m-1) {
            return cost;
        }
        
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
            
            if (maps[nx][ny] == 1) {
                maps[nx][ny] = cost + 1;
                queue.push_back({nx, ny, cost + 1});
            }
        }
    }
    
    return -1;
}